weavetest/abci.go

```diff
@@ -14,6 +14,20 @@ import (
 	abci "github.com/tendermint/tendermint/abci/types"
 )
 
+// Strategy defines which functions we call in ProcessAllTxs
+type Strategy int
+
+const (
+	// CheckOnly will only run Check
+	CheckOnly Strategy = iota
+	// DeliverOnly will only run Deliver
+	DeliverOnly
+	// CheckAndDeliver will call Check for all txs in a block, then Deliver for all txs
+	CheckAndDeliver
+	// DeliverWithPrecheck will call Check *with timer stopped* then time Deliver for all txs
+	DeliverWithPrecheck
```

Comment:
I think this would be better using a bitmask. It would require only 3 values, one for each function to call and could be used like this:
```go
// check and deliver
ProcessAllTxs(Check|Deliver)

// deliver with precheck
ProcessAllTxs(Deliver|Precheck)
```
This is also easier to support inside of the `ProcessAllTxs`.

Standard library `log` package is [using this for flags as well](https://golang.org/pkg/log/#pkg-constants).


We can create a custom bitmask type to make it easier to work with for us
```go
type Bitmask uint32
func (f Bitmask) Has(flag Bitmask) bool { return f&flag != 0 }
func (f *Bitmask) Add(flag Bitmask)     { *f |= flag }
func (f *Bitmask) Remove(flag Bitmask)   { *f &= ^flag }
func (f *Bitmask) Toggle(flag Bitmask)  { *f ^= flag }
```

Reply:
Sounds Good. 

Do you want to make that change and push to this branch?

Reply:
After working on this, I think this can be simplified greatly. I think there should be just two cases:
1. check only
2. deliver only
No other cases (check and deliver or check without measuring and deliver) are required. When writing a tests/benchmark, we can take shortcuts. 

The simple case is we only want to test/benchmark `CheckTx` call. This is simple.

Second case is we want to test/benchmark `DeliverTx`. Weave application always calls `CheckTx` first, but this is not needed for tests. We always assume that check/deliver will succeed in benchmarks. No need to call check when testing delivery - there is a separate test for the check already.

@ethanfrey what do you think?

Reply:
There may be interactions between check and deliver. Like with large sets, loading the data from disk during Check may make subsequent interactions with Deliver faster.

I would like to keep all cases. Not test them all every time, but as you see, there is a series of parameter changes (tx per block) with the others constant, then a series of changes to strategy on one particular setting (with fees, 100tx). We don't need a full matrix, but there are some future optimizations that may take advantage of this more realistic scenarios.

In particular, https://github.com/iov-one/weave/issues/16 is designed to take advantage of the case where Check is run previously to the block processing (during the consensus run, therefore not adding a sequential delay to block production) and we can make use of that to speed up Deliver (which is sequentially blocking the block production), since we don't start another consensus round until the app hash of the last block is known.

Reply:
I would prefer to keep all options there, even if 90% of the test cases are only `Check` or `Deliver`

Reply:
I have extracted configuration to `ProcessOptions` and made it kind of functional configuration. I do not like this solution and I am going to try something else instead.

Here is the code https://github.com/iov-one/weave/commit/a0b0acd8e1d1cedd37e06240fe78707873a2f9ba I will delete it in a day or two :wastebasket: